### PR TITLE
Feature/149 send text messages

### PR DESCRIPTION
**Pull request by timohuber**

### CODE REVIEW

File: pool/test/session_test.ml [hunk 1]
```diff
@@ -511,6 +511,10 @@ let create_cancellation_message reason contact =
   |> CCResult.return
 ;;
 
+let create_cancellation_text_message _ _ phone_number =
```
Comment (mabiede) [inline pool/test/session_test.ml:514]: Which arguments are skipped here, add types

File: pool/test/session_test.ml [hunk 2]
```diff
@@ -523,9 +527,13 @@ let cancel_no_reason () =
   in
   let res =
     SessionC.Cancel.(
-      [ "reason", [ "" ]; "email", [ "true" ]; "sms", [ "true" ] ]
+      [ "reason", [ "" ]; "notify_via", [ "email" ] ]
```
Comment (mabiede) [inline pool/test/session_test.ml:530]: Is the requirement to notify on a single channel, or do we need the possibility to notify on more than one at a time (e.g. email and SMS together)?

File: pool/web/handler/contact_user_profile.ml [hunk 3]
```diff
@@ -147,6 +161,152 @@ let update_password req =
...
+let update_phone_number req =
+  let%lwt urlencoded = Sihl.Web.Request.to_urlencoded req in
+  let result
+    ({ Pool_context.database_label; language; query_language; _ } as context)
+    =
+    let open Utils.Lwt_result.Infix in
+    let tags = tags req in
+    Utils.Lwt_result.map_error (fun msg ->
+      HttpUtils.(
+        msg, "/user/contact-information", [ urlencoded_to_flash urlencoded ]))
```
Comment (mabiede) [inline pool/web/handler/contact_user_profile.ml:173]: `"/user/contact-information"` is used in many places bellow, extract to a variable

File: pool/web/view/component/component_notification.ml [hunk 4]
```diff
@@ -7,6 +7,7 @@ let style_to_class = function
   | `Warning -> "warning"
 ;;
 
+(* TODO[timhub]: Why is the link content hardcoded? *)
```
Comment (mabiede) [inline pool/web/view/component/component_notification.ml:10]: fix todo ;-)

- may use a `?(link : string * I18n.t)` tuple as an optional input argument
- use `CCOption.map_or ~default:(txt "") (fun (path, label) -> ... ) link`

File: pool/web/handler/root_tenant_update.ml [hunk 5]
```diff
@@ -30,6 +30,7 @@ let update req command success_message =
       err, redirect_path, [ HttpUtils.urlencoded_to_flash urlencoded ])
     @@
     let events tenant =
+      Logs.info (fun m -> m "%s" "============ evebts ==========");
```
Comment (mabiede) [inline pool/web/handler/root_tenant_update.ml:33]: left over? ;-)

File: pool/web/view/page/page_root_tenant.ml [hunk 6]
```diff
@@ -97,6 +99,13 @@ let tenant_form
         ~value:(value (fun t -> t.url |> Url.value))
         ~flash_fetcher
         ~required:true
+    ; input_element
+        language
+        `Text
+        Field.GtxApiKey
+        (* ~value:(value (fun t -> t.gtx_api_key |> map_or GtxApiKey.value)) *)
```
Comment (mabiede) [inline pool/web/view/page/page_root_tenant.ml:106]: commented code

File: pool/web/view/page/page_root_tenant.ml [hunk 7]
```diff
@@ -21,6 +21,10 @@ let database_fields tenant language flash_fetcher =
   |> div ~a:[ a_class [ "stack" ] ]
 ;;
 
+let map_or (fnc : 'a -> string) (x : 'a option) =
+  CCOption.map_or ~default:"" fnc x
```
Comment (mabiede) [inline pool/web/view/page/page_root_tenant.ml:25]: Shouldn't this be possible without specifying the input arguments?

```suggestion
let map_or = CCOption.map_or ~default:""
```

File: pool/web/view/page/page_contact_experiment.ml [hunk 8]
```diff
@@ -189,24 +183,44 @@ let show
...
+        (experiment.Public.id |> Experiment.Id.value)
+      |> (fun url ->
+           if user_is_enlisted then Format.asprintf "%s/remove" url else url)
+      |> Sihl.Web.externalize_path
+    in
+    let text_blocks =
+      let base =
+        if user_is_enlisted
+        then p [ txt (hint_to_string I18n.ContactOnWaitingList) ]
+        else p [ txt (hint_to_string I18n.SignUpForWaitingList) ]
```
Comment (mabiede) [inline pool/web/view/page/page_contact_experiment.ml:203]:
```suggestion
        (if user_is_enlisted
        then I18n.ContactOnWaitingList 
        else I18n.SignUpForWaitingList)
        |> fun msg -> p [ txt (hint_to_string msg) ]
```

File: pool/web/view/page/page_contact_experiment.ml [hunk 9]
```diff
@@ -189,24 +183,44 @@ let show
...
+      in
+      let missing_phone =
```
Comment (mabiede) [inline pool/web/view/page/page_contact_experiment.ml:205]: May handle `missing_phone` as an array and use list concat for the combined div. So we don't need the empty `txt`

File: pool/web/view/page/page_admin_session.ml [hunk 10]
```diff
@@ -1011,6 +1011,9 @@ let cancel
       (Experiment.Id.value experiment.Experiment.id)
       Session.(Id.value session.id)
   in
+  let field_to_string f =
+    f |> Utils.field_to_string language |> CCString.capitalize_ascii |> txt
+  in
```
Comment (mabiede) [inline pool/web/view/page/page_admin_session.ml:1016]:
```suggestion
  let field_to_string = Utils.field_to_string language %> CCString.capitalize_ascii %> txt in
```

**Review by mabiede — CHANGES_REQUESTED** [review]

File: pool/app/email/email.mli [hunk 11]
```diff
@@ -1,49 +1,3 @@
-type email_layout =
-  { link : string
-  ; logo_src : string
-  ; logo_alt : string
-  }
-
-val equal_email_layout : email_layout -> email_layout -> bool
-
-module CustomTemplate : sig
```
[the remark this answers is not in the document]
Comment (timohuber) [reply pool/app/email/email.mli:9]: Yes, that was a leftover.

File: pool/app/text_message/text_message_service.ml [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+Text:
+%s
+-----------------------
+    |}
+    (Pool_tenant.Title.value sender)
+    (PhoneNumber.value recipient)
+    text
+;;
+
+let print_message ~tags ?(log_level = Logs.Info) msg =
```
[the remark this answers is not in the document]
Comment (timohuber) [reply pool/app/text_message/text_message_service.ml:75]: So I don't have to return an `Lwt.t` in the handle function manually ;)

File: pool/test/session_test.ml (hunk 2 above)
Comment (timohuber) [reply pool/test/session_test.ml:530]: That is a good question. I will check.

File: pool/web/view/component/component_notification.ml (hunk 4 above)
Comment (timohuber) [reply pool/web/view/component/component_notification.ml:10]: I guess that was the idea :)

**Review by mabiede — APPROVED** [review]

**Pull request merged**